### ISSUE TITLE
Feature Request: Refactory SDAnimatedImageView with the player (frame-based engine) and view (rendering engine), make it possible for custom external usage (SwiftUI/WatchKit)

### ISSUE THREAD

**Issue by dreampiggy** [SDWebImage/SDWebImage#2861]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Feature Request

This is a PR to refactory current `SDAnimatedImage`, seperate the player logic code and view rendering code into two parts. Make it external usage to use player without binding of that UIView subclassing.

### Reason

Our [SDAnimatedImageView solution](https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#animated-image-50) works well on current implementation, a drop-in replacement for UIImageView to play animation just in usage, with frame cache size, progressive loading, everything it's great.

However, current it place the player engine (Which produce image frame and wait time duration), view rendering (Layer contents) code into once place. Sometimes, we don't want or even have a UIView representation, but we still want to use that player to provide CGImage based on time.

For now, our animation solution, can not support WatchKit (because `WKInterfaceObject` can not been created dynamically in code, no subclassing). Our SwiftUI port of SDWebImage's `AnimatedImage` also has some limit because of UIViewRepresentable environment.

I want to make it more abstract and benefit to external usage from all the developers, even you don't use the `SDAnimatedImageView`, you can still build your own animation based on any future context.

This is just like a [AVPlayer](https://developer.apple.com/documentation/avfoundation/avplayer?language=objc) and [AVPlayerLayer](https://developer.apple.com/documentation/avfoundation/avplayerlayer?language=objc) relationship. Apple already use the design like this.

### Draft Implementation

For player part, there is one simple interface draft:

```objective-c
@interface SDAnimatedImagePlayer : NSObject

- (instancetype)initWithProvider:(id<SDAnimatedImageProvider>)provider;
+ (instancetype)playerWithProvider:(id<SDAnimatedImageProvider>)provider;

@property (readonly) UIImage *currentFrame;
@property (readonly) NSUInteger currentIndex;
@property (readonly) NSUInteger currentLoopCount;

#pragma mark - Callback

@property (nonatomic, copy) void (^onFrameChange)(NSUInteger index, UIImage *frame);
@property (nonatomic, copy) void (^onLoopEnd)(NSUInteger loopCount);
 
#pragma mark - Animation Control

@property (readonly) BOOL isPlaying;
- (void)startPlaying;
- (void)pausePlaying;
- (void)stopPlaying;

#pragma mark - Progressive Playing

- (void)updateProvider:(id<SDAnimatedImageProvider>)provider;
- (void)configIndex:(NSUIntger)index; // seek to index

@end
```

You can supply any `SDAnimatedImage` (Since it conforms to `SDAnimatedImageProvider` protocol) to play. The player will callback your onFrameChange handler when the next frame time is reached. This interface also looks like this [CGAnimateImageDataWithBlock](https://developer.apple.com/documentation/imageio/3333272-cganimateimagedatawithblock?language=objc) API from Image/IO

### For SDAnimatedImageView

For current `SDAnimatedImageVierw`, the change is to just grab the frame, pick the CGImage and set to the layer.contents. It's so easy to maintain without understand the details. See like this:

```objective-c
@implementation SDAnimatedImageView

- (void)setImage:(UIImage *)image {
    //...
    self.player.onFrameChange = ^(NSUInteger index, UIImage *frame){
        self.currentFrame = frame;
        self.currentFrameIndex = index;
        [self.imageViewLayer setNeedsDisplay];
    }
    self.player.onLoopEnd = ^(NSUInteger loopCount) {
        self.currentLoopCount = loopCount;
    }
}

@end
```

### For Progressive Animation

Current `SDAnimatedImageView` progressive animation can still been supported. Just need to pause the player (`player.pausePlaying`), then update new data available (`player.updateProvider:newImage`), restore the previous frame index (`player.configIndex:previousIndex`). It can been done in several line of code. And any external usage can built their own progressive animation support as well.

### For SwiftUI Animation

For future the [SDWebImageSwiftUI](https://github.com/SDWebImage/SDWebImageSwiftUI) `AnimatedImage` solution, we can just use the Swift native View and data binding to play animation. No more need `UIViewRepresentable` and support watchOS as well (previouslly it's impossible because WatchKit interface object does not have CALayer :))

```swift
public struct AnimatedImage {
    @State currentIndex: UInt // When this value changed, SwiftUI will trigger body method for new view
    var currentFrame: PlatformImage? // UIImage/NSImage
    var player: SDAnimatedImagePlayer

    public init(data: Data)? {
        guard let image = SDAnimatedImage(data: data) else { return }
        player = SDAnimatedImagePlayer(image)
        player.onFrameChange = [weak self] (index, frame) in {
            guard let self = self else { return }
            self.currentFrame = frame
            self.currentIndex = index
        }
    }

    public var body: some View {
        let image: Image
        if let currentFrame = currentFrame {
            image = Image(platformImage: currentFrame)
        } else {
            image = Image(platformImage: .empty)
        }
        return image
        .onAppear {
            self.player.startPlaying()
        }
        .onDisappear {
            self.player.stopPlaying()
        }
    }
}
```

### For WatchKit Animation

For the legacy WatchKit, since you can not subclassing any `WKInterfaceObject`, so we can bind the logic to `WKInterfaceController` with a Storyboard-linked `WKInterfaceImage`. Just as simple as you think:

```objectivec
@interface MyInterfaceController : WKInterfaceController
@property (weak) IBOutlet WKInterfaceImage *imageInterface;

- (void)awakeWithContext {
    [super awakeWithContext];
    NSData *data = context; // Anyway the animated image data comes from
    SDAnimatedImage *image = [SDAnimatedImage imageWithData:data];
    self.player = [SDAnimatedImagePlayer playerWithProvider:image];
    self.player.onFrameChange = ^(NSUInteger index, UIImage *frame){
        [self.imageInterface setImage:frame];
    }
    self.player.onLoopEnd = ^(NSUInteger loopCount) {
        self.currentLoopCount = loopCount;
    }
}

- (void)didAppear {
    [super didAppear];
    [self.player startPlaying];
}

- (void)willDisappear {
    [super willDisappear];
    [self.player stopPlaying];
}

@end
```

**Comment by dreampiggy** [comment]
Extra notes on implementation (not interface):

1. For the timer solution, we use `CADisplayLink` on UIKit, `CVDisplayLink` on AppKit. However, this class supports watchOS as well. Both of them is unavailable on watchOS. A solution is to use `Dispatch Source Timer` or `NSTimer` instead
2. The player's implementation can be thread-safe. Therefore we can make it possible for more context, like Animated Image Transforming (if possible in future ?)
3. All current `SDAnimatedImageView` public API should not been changed, the configuration property like `maxBufferSize` should be mapped to new player API.

**Comment by dreampiggy** [comment]
Implemented by #2882
